src/CodeGen_PyTorch.cpp

```diff
@@ -0,0 +1,395 @@
+      << print_name(buffer_args[i].name) 
+      << ");\n";
+    if(isCuda) {
+      do_indent();
+      stream
+        << "HLPT_CHECK_DEVICE("
+        << print_name(buffer_args[i].name) 
+        << ", device_id);\n";
+    }
+  }
+  stream << "\n";
+
+  do_indent();
+  stream << "// Wrap tensors in Halide buffers\n";
+  for (size_t i = 0; i < buffer_args.size(); i++) {
+    if (!buffer_args[i].is_buffer())
+      continue;
+
+    do_indent();
+    string tp = type_to_c_type(buffer_args[i].type, false);
+    stream
+      << "Buffer<" << tp << "> "
+      << print_name(buffer_args[i].name) 
+      << "_buffer = Halide::Pytorch::wrap<" << tp << ">("
+      << print_name(buffer_args[i].name) 
+      << ");\n"
+      ;
+  }
+  stream << "\n";
+
+  do_indent();
+  stream << "// Run Halide pipeline\n";
+  do_indent();
+  stream << "int err = " << simple_name << "(";
+  for (size_t i = 0; i < args.size(); i++) {
+    if (args[i].is_buffer()) {
+      stream 
+        << print_name(args[i].name)
+        << "_buffer";
+    } else {
+      stream << print_name(args[i].name);
+    }
+    if (i < args.size()-1) stream << ", ";
+  }
+  stream << ");\n";
+  do_indent();
+  stream << "AT_ASSERTM(err == 0, \"Halide call failed\");\n";
+
+  if(isCuda) {
+    do_indent();
+    stream << "// Make sure data is on device\n";
+    for (size_t i = 0; i < buffer_args.size(); i++) {
+      if (buffer_args[i].is_buffer()) {
+        do_indent();
+        stream 
+          << "AT_ASSERTM(!"
+          << print_name(buffer_args[i].name) << "_buffer.host_dirty(),"
+          << "\"device not synchronized for buffer "
+          << print_name(buffer_args[i].name)
+          << ", make sure all update stages are excplicitly computed on GPU."
+          <<"\");\n";
+        do_indent();
+        stream 
+          << print_name(buffer_args[i].name) << "_buffer"
+          << ".device_detach_native();\n";
+      }
+    }
+    stream << "\n";
+  }
+
+  // do_indent();
+  // stream << "// Free references\n";
+  // for (size_t i = 0; i < buffer_args.size(); i++) {
+  //   if (buffer_args[i].is_buffer()) {
+  //     do_indent();
+  //     stream
+  //       << type_to_pytorch_tensor(buffer_args[i].type, isCuda)
+  //       << "_free(";
+  //     if(isCuda) {
+  //       stream << "state, ";
+  //     }
+  //     stream
+  //       << print_name(buffer_args[i].name) 
+  //       << ");\n"
+  //       ;
+  //   }
+  // }
+  // stream << "\n";
+
+  // TODO(mgharbi): this is not very well documented
+  if (get_env_variable("FLUSH_MEMOIZE_CACHE") == "1") {
+      do_indent();
+      stream << "// Flush cache\n";
+      if (isCuda) {
+          do_indent();
+          stream << "halide_memoization_cache_cleanup(__user_context);\n";
+      } else {
+          do_indent();
+          stream << "halide_memoization_cache_cleanup(NULL);\n";
+      }
+  }
+
+  do_indent();
+  stream << "return 0;\n";
+
+  indent -= 2;
+  stream << "}\n\n";
+
+  // // Pybind interface
```

Comment:
Why is this commented out?

Reply:
removed, we now synthesize the pybind code on the python side (see apps/HelloPyTorch/setup.py)